File: ruby/03_parent_child_calls.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Recipe 03: parent + child Calls (RAG-shaped trace).
#
# Demonstrates Trace structure: one parent Call with two child Calls
# underneath. Children declare their parent via `parent_id` on
# /call/start and share the parent's `trace_id` explicitly.
#
# The RAG-shaped flow:
#     rag_pipeline (parent)
#     ├── retrieve  (child 1)
#     └── generate  (child 2)
#
# Ordering matters: a child's /call/start happens after the parent's
# /call/start, and each child's /call/end happens before the parent's
# /call/end. The recipe shows this canonical order.
#
# Run:
#   ruby ruby/03_parent_child_calls.rb

require "json"
require "net/http"
require "time"
require "uri"

BASE_URL = ENV.fetch("WEAVE_SERVICE_URL", "https://trace.wandb.ai")

required = %w[WANDB_API_KEY WANDB_ENTITY WANDB_PROJECT]
missing = required.reject { |k| ENV[k] && !ENV[k].empty? }
abort "Missing required env vars: #{missing.join(", ")}. See ../README.md#setup." unless missing.empty?

PROJECT_ID = "#{ENV.fetch("WANDB_ENTITY")}/#{ENV.fetch("WANDB_PROJECT")}"
API_KEY = ENV.fetch("WANDB_API_KEY")

BASE_ATTRIBUTES = {
  "cookbook.language" => "ruby",
  "cookbook.recipe" => "03_parent_child_calls",
  "cookbook.environment" => ENV.fetch("COOKBOOK_ENVIRONMENT", "dev"),
}.freeze

def post_json(path, body)
  uri = URI.join(BASE_URL, path)
  req = Net::HTTP::Post.new(uri, "Content-Type" => "application/json")
  req.basic_auth("api", API_KEY)
  req.body = JSON.dump(body)
  res = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |http| http.request(req) }
  abort "HTTP #{res.code} for #{path}: #{res.body}" unless res.code.start_with?("2")
  res.body.empty? ? {} : JSON.parse(res.body)
end

def stream_query(body)
  uri = URI.join(BASE_URL, "/calls/stream_query")
  req = Net::HTTP::Post.new(uri, "Content-Type" => "application/json")
  req.basic_auth("api", API_KEY)
  req.body = JSON.dump(body)

  buffer = +""
  rows = []
  Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
    http.request(req) do |res|
      abort "HTTP #{res.code} for /calls/stream_query: #{res.read_body}" unless res.code.start_with?("2")
      res.read_body do |chunk|
        buffer << chunk
        while (idx = buffer.index("\n"))
          line = buffer.slice!(0..idx).chomp
          rows << JSON.parse(line) unless line.empty?
        end
      end
    end
  end
  rows << JSON.parse(buffer) unless buffer.strip.empty?
  rows
end

def start_call(op_name, inputs, parent_id: nil, trace_id: nil)
  payload = {
    "start" => {
      "project_id" => PROJECT_ID,
      "op_name" => op_name,
      "started_at" => Time.now.utc.iso8601,
      "attributes" => BASE_ATTRIBUTES,
      "inputs" => inputs,
    },
  }
  payload["start"]["parent_id"] = parent_id if parent_id
  payload["start"]["trace_id"] = trace_id if trace_id
  post_json("/call/start", payload)
end

def end_call(call_id, output)
  post_json("/call/end", {
    "end" => {
      "project_id" => PROJECT_ID,
      "id" => call_id,
      "ended_at" => Time.now.utc.iso8601,
      "summary" => {},
      "output" => output,
    },
  })
end

# Open the parent (top-level: no parent_id, no explicit trace_id).
# The server assigns a trace_id which we propagate to children.
parent = start_call("recipe-03-rag-pipeline", { "question" => "Where is the Eiffel Tower?" })
parent_id = parent.fetch("id")
trace_id = parent.fetch("trace_id")
puts "Started parent: id=#{parent_id} trace_id=#{trace_id}"

# Open + finish the first child (retrieve), passing the parent's id and trace_id.
retrieve = start_call(
  "recipe-03-retrieve",
  { "question" => "Where is the Eiffel Tower?" },
  parent_id: parent_id,
  trace_id: trace_id,
)
retrieve_id = retrieve.fetch("id")
puts "Started child 1: id=#{retrieve_id}"
end_call(retrieve_id, { "docs" => ["Paris", "France"] })
puts "Ended   child 1: id=#{retrieve_id}"

# Open + finish the second child (generate).
generate = start_call(
  "recipe-03-generate",
  { "docs" => ["Paris", "France"], "question" => "Where is the Eiffel Tower?" },
  parent_id: parent_id,
  trace_id: trace_id,
)
generate_id = generate.fetch("id")
puts "Started child 2: id=#{generate_id}"
end_call(generate_id, { "answer" => "In Paris, France." })
puts "Ended   child 2: id=#{generate_id}"

# Close the parent (after all children have finished).
end_call(parent_id, { "answer" => "In Paris, France." })
puts "Ended   parent:  id=#{parent_id}"

# --- verification ---
expected = [parent_id, retrieve_id, generate_id]
found_by_id = {}
5.times do
  rows = stream_query({
    "project_id" => PROJECT_ID,
    "filter" => { "trace_ids" => [trace_id] },
  })
  found_by_id = rows.each_with_object({}) { |c, h| h[c["id"]] = c }
  # Require all three visible AND finalized (ended_at populated) so we
  # don't race write-to-read propagation on inner-field reads.
  break if (expected - found_by_id.keys).empty? && expected.all? { |i| found_by_id[i]["ended_at"] }

  sleep 1
end

missing_ids = expected - found_by_id.keys
abort "FAIL: trace #{trace_id} missing calls: #{missing_ids.inspect}" unless missing_ids.empty?

parent_call = found_by_id[parent_id]
retrieve_call = found_by_id[retrieve_id]
generate_call = found_by_id[generate_id]

abort "parent has parent_id: #{parent_call["parent_id"].inspect}" unless parent_call["parent_id"].nil?
abort "retrieve.parent_id: #{retrieve_call["parent_id"].inspect}" unless retrieve_call["parent_id"] == parent_id
abort "generate.parent_id: #{generate_call["parent_id"].inspect}" unless generate_call["parent_id"] == parent_id

[parent_call, retrieve_call, generate_call].each do |c|
  abort "trace_id on #{c["id"]}: #{c["trace_id"].inspect}" unless c["trace_id"] == trace_id
  BASE_ATTRIBUTES.each do |k, v|
    abort "attribute #{k} on #{c["id"]}: #{c["attributes"][k].inspect}" unless c["attributes"][k] == v
  end
end

puts "Verified: trace_id=#{trace_id} (1 parent + 2 children)"

File: java/03_ParentChildCalls.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17+
//DEPS com.fasterxml.jackson.core:jackson-databind:2.18.2

// Recipe 03: parent + child Calls (RAG-shaped trace).
//
// Demonstrates Trace structure: one parent Call with two child Calls
// underneath. Children declare their parent via `parent_id` on
// /call/start and share the parent's `trace_id` explicitly.
//
// The RAG-shaped flow:
//   rag_pipeline (parent)
//   |-- retrieve  (child 1)
//   `-- generate  (child 2)
//
// Ordering matters: a child's /call/start happens after the parent's
// /call/start, and each child's /call/end happens before the parent's
// /call/end. The recipe shows this canonical order.
//
// Verification queries /calls/stream_query by trace_id, gets all three
// Calls back, and asserts the parent/child structure is what we wrote.
//
// Run:
//   jbang java/03_ParentChildCalls.java

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class ParentChildCalls {
    static final ObjectMapper MAPPER = new ObjectMapper();
    static final HttpClient HTTP = HttpClient.newHttpClient();
    static final String BASE_URL = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai");
    static String projectId;
    static String authHeader;
    static Map<String, Object> baseAttributes;

    static String getenv(String key, String def) {
        String v = System.getenv(key);
        return (v == null || v.isEmpty()) ? def : v;
    }

    static void fail(String msg) {
        System.err.println("FAIL: " + msg);
        System.exit(1);
    }

    // Centralizes auth + JSON (de)serialization. Any non-2xx response is fatal.
    static JsonNode postJson(String path, Object body) throws Exception {
        String json = MAPPER.writeValueAsString(body);
        HttpRequest req = HttpRequest.newBuilder(URI.create(BASE_URL + path))
                .header("Authorization", authHeader)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
        HttpResponse<String> res = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2)
            fail("HTTP " + res.statusCode() + " for " + path + ": " + res.body());
        String b = res.body();
        return (b == null || b.isEmpty()) ? MAPPER.createObjectNode() : MAPPER.readTree(b);
    }

    // Parses the NDJSON /calls/stream_query response line-by-line.
    static List<JsonNode> streamQuery(Object body) throws Exception {
        String json = MAPPER.writeValueAsString(body);
        HttpRequest req = HttpRequest.newBuilder(URI.create(BASE_URL + "/calls/stream_query"))
                .header("Authorization", authHeader)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
        HttpResponse<Stream<String>> res = HTTP.send(req, HttpResponse.BodyHandlers.ofLines());
        if (res.statusCode() / 100 != 2)
            fail("HTTP " + res.statusCode() + " for /calls/stream_query: "
                    + res.body().collect(Collectors.joining("\n")));
        var rows = new ArrayList<JsonNode>();
        for (Iterator<String> it = res.body().iterator(); it.hasNext(); ) {
            String line = it.next();
            if (line.isBlank()) continue;
            rows.add(MAPPER.readTree(line));
        }
        return rows;
    }

    // POSTs /call/start. parentId and traceId are omitted when null, so a
    // top-level Call passes null for both and the server assigns a trace_id.
    static JsonNode startCall(String opName, Map<String, Object> inputs, String parentId, String traceId)
            throws Exception {
        var start = new LinkedHashMap<String, Object>();
        start.put("project_id", projectId);
        start.put("op_name", opName);
        start.put("started_at", Instant.now().toString());
        start.put("attributes", baseAttributes);
        start.put("inputs", inputs);
        if (parentId != null) start.put("parent_id", parentId);
        if (traceId != null) start.put("trace_id", traceId);
        return postJson("/call/start", Map.of("start", start));
    }

    // POSTs /call/end.
    static void endCall(String callId, Map<String, Object> output) throws Exception {
        postJson("/call/end", Map.of(
                "end", Map.of(
                        "project_id", projectId,
                        "id", callId,
                        "ended_at", Instant.now().toString(),
                        "summary", Map.of(),
                        "output", output)));
    }

    public static void main(String[] args) throws Exception {
        var missing = new ArrayList<String>();
        for (String k : List.of("WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"))
            if (getenv(k, "").isEmpty()) missing.add(k);
        if (!missing.isEmpty())
            fail("Missing required env vars: " + String.join(", ", missing) + ". See ../README.md#setup.");

        projectId = System.getenv("WANDB_ENTITY") + "/" + System.getenv("WANDB_PROJECT");
        authHeader = "Basic " + Base64.getEncoder().encodeToString(
                ("api:" + System.getenv("WANDB_API_KEY")).getBytes(StandardCharsets.UTF_8));
        baseAttributes = new LinkedHashMap<>();
        baseAttributes.put("cookbook.language", "java");
        baseAttributes.put("cookbook.recipe", "03_parent_child_calls");
        baseAttributes.put("cookbook.environment", getenv("COOKBOOK_ENVIRONMENT", "dev"));

        String question = "Where is the Eiffel Tower?";
        List<String> docs = List.of("Paris", "France");
        String answer = "In Paris, France.";

        // Open the parent (top-level: no parent_id, no explicit trace_id).
        // The server assigns a trace_id which we propagate to children.
        JsonNode parent = startCall("recipe-03-rag-pipeline", Map.of("question", question), null, null);
        String parentId = parent.get("id").asText();
        String traceId = parent.get("trace_id").asText();
        System.out.println("Started parent: id=" + parentId + " trace_id=" + traceId);

        // Open + finish the first child (retrieve), under the parent + trace.
        JsonNode retrieve = startCall("recipe-03-retrieve", Map.of("question", question), parentId, traceId);
        String retrieveId = retrieve.get("id").asText();
        System.out.println("Started child 1: id=" + retrieveId);
        endCall(retrieveId, Map.of("docs", docs));
        System.out.println("Ended   child 1: id=" + retrieveId);

        // Open + finish the second child (generate).
        JsonNode generate = startCall("recipe-03-generate", Map.of("docs", docs, "question", question), parentId, traceId);
        String generateId = generate.get("id").asText();
        System.out.println("Started child 2: id=" + generateId);
        endCall(generateId, Map.of("answer", answer));
        System.out.println("Ended   child 2: id=" + generateId);

        // Close the parent (after all children have finished).
        endCall(parentId, Map.of("answer", answer));
        System.out.println("Ended   parent:  id=" + parentId);

        // --- verification ---
        // Stream all Calls in this trace; assert parent + 2 children, with
        // parent.parent_id absent and children.parent_id = parent_id.
        List<String> expected = List.of(parentId, retrieveId, generateId);
        var found = new LinkedHashMap<String, JsonNode>();
        for (int i = 0; i < 5; i++) {
            List<JsonNode> rows = streamQuery(Map.of(
                    "project_id", projectId,
                    "filter", Map.of("trace_ids", List.of(traceId))));
            found.clear();
            for (JsonNode c : rows) {
                String id = c.path("id").asText(null);
                if (id != null) found.put(id, c);
            }
            // Require all three visible AND finalized (ended_at populated) so we
            // don't race write-to-read propagation on inner-field reads.
            boolean ready = expected.stream().allMatch(id -> found.containsKey(id) && found.get(id).hasNonNull("ended_at"));
            if (ready) break;
            Thread.sleep(1000);
        }

        for (String id : expected)
            if (!found.containsKey(id))
                fail("trace " + traceId + " missing call " + id);

        JsonNode parentCall = found.get(parentId), retrieveCall = found.get(retrieveId), generateCall = found.get(generateId);
        if (parentCall.hasNonNull("parent_id"))
            fail("parent has parent_id: " + parentCall.get("parent_id"));
        if (!parentId.equals(retrieveCall.path("parent_id").asText(null)))
            fail("retrieve.parent_id: " + retrieveCall.get("parent_id"));
        if (!parentId.equals(generateCall.path("parent_id").asText(null)))
            fail("generate.parent_id: " + generateCall.get("parent_id"));
        for (JsonNode c : List.of(parentCall, retrieveCall, generateCall)) {
            if (!traceId.equals(c.get("trace_id").asText()))
                fail("trace_id on " + c.get("id") + ": " + c.get("trace_id"));
            for (var e : baseAttributes.entrySet()) {
                JsonNode actual = c.path("attributes").get(e.getKey());
                if (actual == null || !actual.asText().equals(e.getValue()))
                    fail("attribute " + e.getKey() + " on " + c.get("id") + ": " + actual);
            }
        }
        System.out.println("Verified: trace_id=" + traceId + " (1 parent + 2 children)");
    }
}
